File: tools/scale-test/sel_benchmarks.py
#!/usr/bin/env python3
"""Python scale/parity runner for the Lisp benchmark oracle."""

from __future__ import annotations

import argparse
import gc
import json
import struct
import sys
import time
from pathlib import Path
from typing import Any

# Keep the runner directly executable from a source checkout, like the other
# host-specific scale runners.
ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / 'python'))

from sel import Value, compile
from sel.registry import define
from sel.sql import Binding, HybridPlan, Sql
from sel.value import structural_hash

sys.path.insert(0, str(Path(__file__).resolve().parent))
from benchmark_support import fixture_metadata, resolve_path, runtime_metadata, sha256_file, stats


def read_json(path: str | Path, *, parse_float=None):
    with resolve_path(path).open(encoding='utf-8') as stream:
        return json.load(stream, parse_float=parse_float)


def parse_whole(text: str, fallback: int) -> int:
    text = str(text).strip()
    head = text.split('.', 1)[0]
    try:
        return int(head)
    except ValueError:
        return fallback


def register_benchmark_builtins() -> None:
    define(
        'CUSTOM_VIP_SCORE', 2, 2,
        fn=lambda args, ctx: Value.int(
            (100 if args.text(0) == 'PLATINUM' else
             50 if args.text(0) == 'GOLD' else
             25 if args.text(0) == 'SILVER' else 10)
            + (2026 - parse_whole(args.text(1), 2024)) * 5),
    )
    define(
        'HOST_RISK_SCORE', 2, 2,
        fn=lambda args, ctx: Value.int(
            (30 if args.text(0) == 'US' else 10)
            + parse_whole(args.text(1), 0) * 2),
    )


def float32(value: str | int | float) -> float:
    return struct.unpack('<f', struct.pack('<f', float(value)))[0]


def load_context(dataset: dict[str, Any]) -> Value:
    context = Value.none()
    for table, rows in dataset.items():
        context.set(str(table).upper(), Value.from_native(rows))
    customers = context.get('CUSTOMERS')
    if customers is not None:
        shaped_customers = []
        customer_values = (
            customers.storage
            if customers.storage is not None
            else [customer for _, customer in customers.entries()]
        )
        for customer in customer_values:
            latitude = float32(customer.get('latitude').as_text())
            longitude = float32(customer.get('longitude').as_text())
            dx = longitude - 13.404954
            dy = latitude - 52.520008
            distance = (dx * dx + dy * dy) ** 0.5
            if customer.shape is not None and customer.storage is not None:
                keys = list(customer.shape.keys)
                values = list(customer.storage)
                keys.append('dist_berlin')
                values.append(Value.text(f'{distance:.6f}'))
                shaped_customers.append(Value.shaped(keys, values))
            else:
                entries = customer.entries()
                entries.append(('dist_berlin', Value.text(f'{distance:.6f}')))
                shaped_customers.append(Value.from_entries(entries))
        context.set('CUSTOMERS', Value.list(shaped_customers))
    return context


def relation(table: str, fields: dict[str, str]) -> Binding:
    mapped = {}
    text_fields = {'CODE', 'NAME', 'SKU', 'COUNTRY', 'TIER', 'STATUS'}
    for key, column in fields.items():
        mapped[key] = Binding.column(column, table,
                                     'TEXT' if key in text_fields else 'NUM')
    return Binding.relation(table, table, mapped)


def schema(dialect: str) -> dict[str, Binding]:
    categories = relation('categories', {
        'ID': 'id', 'CODE': 'code', 'NAME': 'name', 'VAT_RATE': 'vat_rate',
    })
    products = relation('products', {
        'ID': 'id', 'SKU': 'sku', 'NAME': 'name', 'CATEGORY_ID': 'category_id',
        'PRICE': 'price', 'IS_ACTIVE': 'is_active',
    })
    customer_fields = {
        'ID': Binding.column('id', 'customers', 'NUM'),
        'NAME': Binding.column('name', 'customers', 'TEXT'),
        'TIER': Binding.column('tier', 'customers', 'TEXT'),
        'COUNTRY': Binding.column('country', 'customers', 'TEXT'),
        'CREATED_YEAR': Binding.column('created_year', 'customers', 'NUM'),
        'LATITUDE': Binding.column('latitude', 'customers', 'NUM'),
        'LONGITUDE': Binding.column('longitude', 'customers', 'NUM'),
        'DIST_BERLIN': Binding.raw(
            'ROUND((customers.location <-> point(13.404954, 52.520008))::numeric, 6)',
            'NUM') if dialect == 'postgresql' else Binding.raw(
                'ROUND(ST_Distance(POINT(customers.longitude, customers.latitude), '
                'POINT(13.404954, 52.520008)), 6)', 'NUM'),
    }
    return {
        'CATEGORIES': categories,
        'PRODUCTS': products,
        'CUSTOMERS': Binding.relation('customers', 'customers', customer_fields),
        'ORDERS': relation('orders', {
            'ID': 'id', 'CUSTOMER_ID': 'customer_id', 'STATUS': 'status',
            'DISCOUNT': 'discount', 'ORDER_YEAR': 'order_year',
        }),
        'ORDER_ITEMS': relation('order_items', {
            'ID': 'id', 'ORDER_ID': 'order_id', 'PRODUCT_ID': 'product_id',
            'QUANTITY': 'quantity', 'UNIT_PRICE': 'unit_price',
        }),
    }


def canonical(value: Any) -> Any:
    if isinstance(value, list):
        return [canonical(item) for item in value]
    if isinstance(value, dict):
        return {key: canonical(value[key]) for key in sorted(value)}
    return value


def benchmark_value(value: Value) -> Any:
    value.force()
    if value.size() == 0:
        if value.is_list:
            return []
        if value.kind in (Value.TEXT, Value.BIN, Value.BOOL):
            return value.scalar
        return None
    if value.is_list:
        return [benchmark_value(item) for item in value.values()]
    out = {key: benchmark_value(child) for key, child in value.entries()}
    if value.kind != Value.NONE:
        out['_'] = value.scalar
    return out


def same_value(actual: Any, expected: Any) -> bool:
    return canonical(actual) == canonical(expected)


def representation_counts(value: Value, counts: dict[str, int]) -> None:
    value.force()
    if value.is_list:
        counts['lists'] += 1
    elif value.shape is not None:
        counts['shaped_records'] += 1
    elif value.size() > 0:
        counts['fallback_records'] += 1
    if value.storage is not None:
        for child in value.storage:
            representation_counts(child, counts)
    elif value.children:
        for child in value.children.values():
            representation_counts(child, counts)


def run_and_materialize(program, context: Value) -> tuple[float, float, float, Any]:
    prepared_start = time.perf_counter()
    run_start = time.perf_counter()
    actual = program.run(context)
    run_ms = (time.perf_counter() - run_start) * 1000.0
    materialize_start = time.perf_counter()
    rows = benchmark_value(actual)
    materialize_ms = (time.perf_counter() - materialize_start) * 1000.0
    prepared_ms = (time.perf_counter() - prepared_start) * 1000.0
    return run_ms, materialize_ms, prepared_ms, rows


def plan_sql(plan: HybridPlan) -> str | None:
    return None if plan.sql_statement is None else plan.sql_statement.as_statement('inline')


def check_plan(plan: HybridPlan, expected: dict[str, Any], dialect: str,
               failures: list[str]) -> None:
    sql_key = 'sql_postgres' if dialect == 'postgresql' else 'sql_mariadb'
    if plan_sql(plan) != expected.get(sql_key):
        failures.append(f'{dialect} SQL differs from Lisp reference')
    expected_hybrid = (expected.get('is_hybrid') is True
                       or expected.get('has_continuation') is True)
    if plan.is_hybrid != expected_hybrid:
        failures.append(f'{dialect} hybrid flag differs')
    if plan.pure_sql != (not expected_hybrid):
        failures.append(f'{dialect} pureSql flag differs')
    if (plan.continuation_program is not None) != (expected.get('has_continuation') is True):
        failures.append(f'{dialect} continuation presence differs')


def main() -> int:
    parser = argparse.ArgumentParser(description='SEL Python scale/parity runner')
    parser.add_argument('--dataset', default='tools/scale-test/dataset-10x.json')
    parser.add_argument('--reference', default='tools/scale-test/benchmark_results.json')
    parser.add_argument('--only', default=None)
    parser.add_argument('--output', default=None)
    parser.add_argument('--runs', type=int, default=1)
    parser.add_argument('--warmups', type=int, default=0)
    parser.add_argument('--timing-mode', default='steady-state',
                        choices=('steady-state', 'gc-controlled'))
    args = parser.parse_args()
    if args.runs < 1 or args.warmups < 0:
        raise SystemExit('--runs must be positive and --warmups must be non-negative')

    register_benchmark_builtins()
    dataset_path = resolve_path(args.dataset)
    reference_path = resolve_path(args.reference)
    dataset = read_json(dataset_path, parse_float=str)
    reference = read_json(reference_path)
    if not isinstance(reference, list):
        raise SystemExit('reference must be a scenario list')
    reference_ids = [str(item['id']) for item in reference]
    if len(reference_ids) != len(set(reference_ids)):
        raise SystemExit('reference contains duplicate scenario ids')
    selected = reference if args.only is None else [
        item for item in reference if item['id'] in set(args.only.split(','))
    ]
    if not selected:
        raise SystemExit('--only selected no scenarios')
    context = load_context(dataset)
    fixture = fixture_metadata(dataset_path, dataset)
    counts = {'shaped_records': 0, 'fallback_records': 0, 'lists': 0}
    representation_counts(context, counts)
    context_signature = structural_hash(context)

    compile_ms: dict[str, float] = {}
    programs = {}
    plans = {}
    for expected in selected:
        started = time.perf_counter()
        programs[expected['id']] = compile(expected['query'])
        compile_ms[expected['id']] = (time.perf_counter() - started) * 1000.0
        failures: list[str] = []
        plans[expected['id']] = {
            dialect: Sql.plan_hybrid(programs[expected['id']], dialect, schema(dialect))
            for dialect in ('postgresql', 'mariadb')
        }
        for dialect, plan in plans[expected['id']].items():
            check_plan(plan, expected, dialect, failures)
        if failures:
            raise RuntimeError(f"python SQL parity failed for {expected['id']}: {'; '.join(failures)}")

    report = {
        'schema_version': 2,
        'implementation': 'python',
        'passed': True,
        'metadata': {
            'fixture': fixture,
            'reference_path': str(reference_path.resolve()),
            'reference_sha256': sha256_file(reference_path),
            'reference_scenario_ids': [item['id'] for item in selected],
            'runtime': runtime_metadata(),
            'representation': counts,
            'scenario_order': [item['id'] for item in selected],
            'timing_mode': args.timing_mode,
            'gc_policy': 'gc.collect() before warmups and samples' if args.timing_mode == 'gc-controlled' else 'not forced',
            'runs': args.runs,
            'warmups': args.warmups,
        },
        'scenarios': [],
    }

    for expected in selected:
        scenario_id = expected['id']
        failures: list[str] = []
        # The validation execution is intentionally outside all reported phase
        # timers. It proves that this read-only fixture can be reused safely.
        before = structural_hash(context)
        _, _, _, validation_rows = run_and_materialize(programs[scenario_id], context)
        after = structural_hash(context)
        context_unchanged = before == after == context_signature
        if not context_unchanged:
            failures.append('prepared context changed during untimed validation')
        if not same_value(validation_rows, expected['in_memory_rows']):
            failures.append('in-memory rows differ from Lisp reference during validation')

        for warmup in range(args.warmups):
            if args.timing_mode == 'gc-controlled':
                gc.collect()
            print(f'[python] {scenario_id} warmup {warmup + 1}/{args.warmups}', flush=True)
            _, _, _, warm_rows = run_and_materialize(programs[scenario_id], context)
            if not same_value(warm_rows, expected['in_memory_rows']):
                failures.append(f'warmup {warmup + 1} result differs from Lisp reference')
        samples: list[dict[str, float]] = []
        for run in range(args.runs):
            if args.timing_mode == 'gc-controlled':
                gc.collect()
            print(f'[python] {scenario_id} measured {run + 1}/{args.runs}', flush=True)
            run_ms, materialize_ms, prepared_ms, actual_rows = run_and_materialize(
                programs[scenario_id], context)
            if not same_value(actual_rows, expected['in_memory_rows']):
                failures.append(f'measured run {run + 1} result differs from Lisp reference')
            if structural_hash(context) != context_signature:
                failures.append(f'measured run {run + 1} changed prepared context')
            samples.append({
                'program_run_ms': run_ms,
                'materialize_ms': materialize_ms,
                'prepared_total_ms': prepared_ms,
                'elapsed_ms': prepared_ms,
            })
        phase_stats = {
            phase: stats([sample[phase] for sample in samples])
            for phase in ('program_run_ms', 'materialize_ms', 'prepared_total_ms')
        }
        passed = not failures
        report['passed'] = report['passed'] and passed
        report['scenarios'].append({
            'id': scenario_id,
            'rows': len(validation_rows) if isinstance(validation_rows, list) else None,
            'compile_ms': compile_ms[scenario_id],
            'samples': samples,
            'statistics': phase_stats,
            'parity': {'passed': passed, 'failures': failures},
            'context_unchanged': context_unchanged,
            'passed': passed,
            'failures': failures,
        })

    if args.output:
        output = resolve_path(args.output)
        output.write_text(json.dumps(report, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')
    print(f"Python scale parity: {sum(item['passed'] for item in report['scenarios'])}/"
          f"{len(report['scenarios'])} passed; mode={args.timing_mode}; runs={args.runs}; warmups={args.warmups}")
    return 0 if report['passed'] else 1


if __name__ == '__main__':
    raise SystemExit(main())

File: tools/scale-test/sel_benchmarks.mjs
#!/usr/bin/env node
// JavaScript scale/parity runner.
//
// benchmark_results.json is produced by sel_benchmarks.lisp and is the Lisp
// reference for this lane.  This runner deliberately compares both the
// in-memory result and the two generated SQL statements against that oracle;
// database execution remains the responsibility of run_benchmarks.py.

import fs from 'node:fs';
import crypto from 'node:crypto';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { performance } from 'node:perf_hooks';

import {
  Value,
  compile,
  register,
} from '../../js/src/sel.mjs';
import { structuralHash } from '../../js/src/value.mjs';
import { Binding, planHybrid } from '../../js/src/sql/index.mjs';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');

function option(name, fallback) {
  const index = process.argv.indexOf(name);
  return index < 0 ? fallback : process.argv[index + 1] ?? fallback;
}

function readJson(file) {
  return JSON.parse(fs.readFileSync(path.resolve(ROOT, file), 'utf8'));
}

function resolvePath(file) {
  return path.resolve(ROOT, file);
}

function sha256File(file) {
  return crypto.createHash('sha256').update(fs.readFileSync(resolvePath(file))).digest('hex');
}

function fixtureMetadata(file, dataset) {
  const resolved = resolvePath(file);
  const sha256 = crypto.createHash('sha256').update(fs.readFileSync(resolved)).digest('hex');
  const tableRows = Object.fromEntries(Object.entries(dataset).map(([name, rows]) => [
    name.toUpperCase(), Array.isArray(rows) ? rows.length : null,
  ]));
  return {
    path: resolved,
    sha256,
    table_rows: tableRows,
    total_source_rows: Object.values(tableRows).filter((count) => count !== null)
      .reduce((total, count) => total + count, 0),
    schema_version: 1,
  };
}

function runtimeMetadata() {
  return {
    os: `${os.type()} ${os.release()}`,
    machine: os.arch(),
    cpu: os.cpus()[0]?.model ?? 'unknown',
    logical_cpus: os.cpus().length,
    available_memory_bytes: os.freemem(),
    node: process.version,
    v8: process.versions.v8,
    exec_argv: process.execArgv,
  };
}

function phaseStats(samples) {
  if (samples.length === 0) throw new Error('cannot summarize empty samples');
  const values = samples.map(Number);
  if (values.some((value) => !Number.isFinite(value) || value < 0)) {
    throw new Error(`invalid benchmark sample: ${values}`);
  }
  const sorted = [...values].sort((a, b) => a - b);
  const mean = values.reduce((sum, value) => sum + value, 0) / values.length;
  const median = sorted.length % 2 === 1
    ? sorted[(sorted.length - 1) / 2]
    : (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2;
  const variance = values.length > 1
    ? values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / (values.length - 1)
    : 0;
  const rank = Math.max(1, Math.ceil(0.95 * values.length));
  return {
    count: values.length,
    mean_ms: mean,
    median_ms: median,
    min_ms: sorted[0],
    max_ms: sorted[sorted.length - 1],
    stdev_ms: Math.sqrt(variance),
    cv: mean === 0 ? 0 : Math.sqrt(variance) / mean,
    p95_ms: values.length >= 20 ? sorted[rank - 1] : null,
    samples_ms: values,
  };
}

function parseWhole(text, fallback) {
  const value = Number.parseInt(String(text).split('.')[0], 10);
  return Number.isFinite(value) ? value : fallback;
}

function registerBenchmarkBuiltins() {
  register({
    name: 'CUSTOM_VIP_SCORE', min: 2, max: 2, overwrite: true,
    fn: (args) => {
      const tier = args.val(0).asText(args.posOf(0));
      const year = parseWhole(args.val(1).asText(args.posOf(1)), 2024);
      const base = tier === 'PLATINUM' ? 100
        : tier === 'GOLD' ? 50
          : tier === 'SILVER' ? 25 : 10;
      return Value.int(base + (2026 - year) * 5);
    },
  });
  register({
    name: 'HOST_RISK_SCORE', min: 2, max: 2, overwrite: true,
    fn: (args) => {
      const country = args.val(0).asText(args.posOf(0));
      const discount = parseWhole(args.val(1).asText(args.posOf(1)), 0);
      return Value.int((country === 'US' ? 30 : 10) + discount * 2);
    },
  });
}

function loadContext(dataset) {
  const context = Value.none();
  for (const [table, rows] of Object.entries(dataset)) {
    context.set(table.toUpperCase(), Value.fromNative(rows));
  }

  // The Lisp fixture uses READ-FROM-STRING, whose default is a single-float
  // for these values.  Math.fround keeps the JS oracle comparison on the same
  // six-decimal boundary without changing SEL's exact decimal arithmetic.
  const customers = context.get('CUSTOMERS');
  if (customers) {
    const shapedCustomers = [];
    const customerValues = customers.storage ?? customers.values();
    for (const customer of customerValues) {
      const latitude = Math.fround(Number(customer.get('latitude').asText()));
      const longitude = Math.fround(Number(customer.get('longitude').asText()));
      const dx = longitude - 13.404954;
      const dy = latitude - 52.520008;
      const distance = Math.sqrt(dx * dx + dy * dy);
      if (customer.shape !== null && customer.storage !== null) {
        const keys = [...customer.shape.keys, 'dist_berlin'];
        const values = [...customer.storage, Value.text(distance.toFixed(6))];
        shapedCustomers.push(Value.shapedOwned(keys, values));
      } else {
        const entries = customer.entries();
        entries.push(['dist_berlin', Value.text(distance.toFixed(6))]);
        shapedCustomers.push(Value.fromEntries(entries));
      }
    }
    context.set('CUSTOMERS', Value.list(shapedCustomers));
  }
  return context;
}

function relation(name, fields) {
  const mapped = {};
  for (const [key, column] of Object.entries(fields)) {
    mapped[key] = Binding.column(column, name, key === 'CODE' || key === 'NAME'
      || key === 'SKU' || key === 'COUNTRY' || key === 'TIER' || key === 'STATUS'
      ? 'TEXT' : 'NUM');
  }
  return Binding.relation(name, name, mapped);
}

function schema(dialect) {
  const categories = relation('categories', {
    ID: 'id', CODE: 'code', NAME: 'name', VAT_RATE: 'vat_rate',
  });
  const products = relation('products', {
    ID: 'id', SKU: 'sku', NAME: 'name', CATEGORY_ID: 'category_id',
    PRICE: 'price', IS_ACTIVE: 'is_active',
  });
  const customersFields = {
    ID: Binding.column('id', 'customers', 'NUM'),
    NAME: Binding.column('name', 'customers', 'TEXT'),
    TIER: Binding.column('tier', 'customers', 'TEXT'),
    COUNTRY: Binding.column('country', 'customers', 'TEXT'),
    CREATED_YEAR: Binding.column('created_year', 'customers', 'NUM'),
    LATITUDE: Binding.column('latitude', 'customers', 'NUM'),
    LONGITUDE: Binding.column('longitude', 'customers', 'NUM'),
    DIST_BERLIN: dialect === 'postgresql'
      ? Binding.raw('ROUND((customers.location <-> point(13.404954, 52.520008))::numeric, 6)', 'NUM')
      : Binding.raw('ROUND(ST_Distance(POINT(customers.longitude, customers.latitude), POINT(13.404954, 52.520008)), 6)', 'NUM'),
  };
  const customers = Binding.relation('customers', 'customers', customersFields);
  const orders = relation('orders', {
    ID: 'id', CUSTOMER_ID: 'customer_id', STATUS: 'status',
    DISCOUNT: 'discount', ORDER_YEAR: 'order_year',
  });
  const orderItems = relation('order_items', {
    ID: 'id', ORDER_ID: 'order_id', PRODUCT_ID: 'product_id',
    QUANTITY: 'quantity', UNIT_PRICE: 'unit_price',
  });
  return {
    CATEGORIES: categories,
    PRODUCTS: products,
    CUSTOMERS: customers,
    ORDERS: orders,
    ORDER_ITEMS: orderItems,
  };
}

function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, canonical(value[key])]));
  }
  return value;
}

function benchmarkValue(value) {
  if (value.size() === 0) {
    if (value.isList) return [];
    if (value.kind === 'TEXT' || value.kind === 'BIN' || value.kind === 'BOOL') return value.scalar;
    return null;
  }
  if (value.isList) return value.values().map(benchmarkValue);
  const out = {};
  for (const [key, child] of value.entries()) out[key] = benchmarkValue(child);
  if (value.kind === 'TEXT' || value.kind === 'BIN' || value.kind === 'BOOL') out._ = value.scalar;
  return out;
}

function representationCounts(value, counts) {
  if (value.isList) counts.lists += 1;
  else if (value.shape) counts.shaped_records += 1;
  else if (value.size() > 0) counts.fallback_records += 1;
  if (value.storage !== null) {
    for (const child of value.storage) representationCounts(child, counts);
  } else if (value.children) {
    for (const child of value.children.values()) representationCounts(child, counts);
  } else if (value._entries) {
    for (const [, child] of value._entries) representationCounts(child, counts);
  }
}

function runAndMaterialize(program, context) {
  const preparedStart = performance.now();
  const runStart = performance.now();
  const actual = program.run(context);
  const runMs = performance.now() - runStart;
  const materializeStart = performance.now();
  const rows = benchmarkValue(actual);
  const materializeMs = performance.now() - materializeStart;
  const preparedMs = performance.now() - preparedStart;
  return { runMs, materializeMs, preparedMs, rows };
}

function equalValues(actual, expected) {
  return JSON.stringify(canonical(actual)) === JSON.stringify(canonical(expected));
}

function planSql(plan) {
  return plan.sqlStatement === null ? null : plan.sqlStatement.asStatement('inline');
}

function checkPlan(plan, expected, dialect, failures) {
  const sqlKey = dialect === 'postgresql' ? 'sql_postgres' : 'sql_mariadb';
  const actualSql = planSql(plan);
  const expectedSql = expected[sqlKey] ?? null;
  if (actualSql !== expectedSql) {
    failures.push(`${dialect} SQL differs from Lisp reference`);
  }
  const expectedHasSql = expectedSql !== null;
  const expectedContinuation = expected.has_continuation === true;
  const expectedHybrid = expectedHasSql && expectedContinuation;
  const expectedPureSql = expectedHasSql && !expectedContinuation;
  if (plan.isHybrid !== expectedHybrid) {
    failures.push(`${dialect} hybrid flag is ${plan.isHybrid}, expected ${expectedHybrid}`);
  }
  if (plan.pureSql !== expectedPureSql) {
    failures.push(`${dialect} pureSql flag is ${plan.pureSql}, expected ${expectedPureSql}`);
  }
  const hasContinuation = plan.continuationProgram !== null;
  if (hasContinuation !== expectedContinuation) {
    failures.push(`${dialect} continuation presence differs from Lisp reference`);
  }
}

function correctedMain() {
  registerBenchmarkBuiltins();
  const datasetFile = option('--dataset', 'tools/scale-test/dataset-10x.json');
  const referenceFile = option('--reference', 'tools/scale-test/benchmark_results.json');
  const dataset = readJson(datasetFile);
  const reference = readJson(referenceFile);
  const only = option('--only', null);
  const runs = parseWhole(option('--runs', '1'), 1);
  const warmups = parseWhole(option('--warmups', '0'), 0);
  const timingMode = option('--timing-mode', 'steady-state');
  if (runs < 1 || warmups < 0) throw new Error('--runs must be positive and --warmups non-negative');
  if (!['steady-state', 'gc-controlled'].includes(timingMode)) {
    throw new Error(`unsupported timing mode: ${timingMode}`);
  }
  if (timingMode === 'gc-controlled' && typeof global.gc !== 'function') {
    throw new Error('gc-controlled mode requires Node.js --expose-gc');
  }
  const selected = only === null
    ? reference
    : reference.filter((scenario) => new Set(only.split(',')).has(scenario.id));
  if (selected.length === 0) throw new Error('--only selected no scenarios');
  const referenceIds = reference.map((scenario) => scenario.id);
  if (new Set(referenceIds).size !== referenceIds.length) throw new Error('reference has duplicate scenario ids');
  const context = loadContext(dataset);
  const representation = { shaped_records: 0, fallback_records: 0, lists: 0 };
  representationCounts(context, representation);
  const contextSignature = structuralHash(context);
  const programs = new Map();
  const compileMs = new Map();

  for (const expected of selected) {
    const compileStart = performance.now();
    const program = compile(expected.query);
    compileMs.set(expected.id, performance.now() - compileStart);
    programs.set(expected.id, program);
    const failures = [];
    for (const dialect of ['postgresql', 'mariadb']) {
      const plan = planHybrid(program, dialect, schema(dialect));
      checkPlan(plan, expected, dialect, failures);
    }
    if (failures.length > 0) throw new Error(`JS SQL parity failed for ${expected.id}: ${failures.join('; ')}`);
  }

  const report = {
    schema_version: 2,
    implementation: 'js',
    scenarios: [],
    passed: true,
    metadata: {
      fixture: fixtureMetadata(datasetFile, dataset),
      reference_path: resolvePath(referenceFile),
      reference_sha256: sha256File(referenceFile),
      reference_scenario_ids: selected.map((scenario) => scenario.id),
      runtime: runtimeMetadata(),
      representation,
      scenario_order: selected.map((scenario) => scenario.id),
      timing_mode: timingMode,
      gc_policy: timingMode === 'gc-controlled' ? 'global.gc() before warmups and samples' : 'not forced',
      gc_available: typeof global.gc === 'function',
      runs,
      warmups,
    },
  };

  for (const expected of selected) {
    const failures = [];
    const program = programs.get(expected.id);
    const validationBefore = structuralHash(context);
    const validation = runAndMaterialize(program, context);
    const contextUnchanged = validationBefore === structuralHash(context)
      && validationBefore === contextSignature;
    if (!contextUnchanged) failures.push('prepared context changed during untimed validation');
    if (!equalValues(validation.rows, expected.in_memory_rows)) {
      failures.push('in-memory rows differ from Lisp reference during validation');
    }
    for (let warmup = 0; warmup < warmups; warmup += 1) {
      if (timingMode === 'gc-controlled') global.gc();
      console.log(`[js] ${expected.id} warmup ${warmup + 1}/${warmups}`);
      const result = runAndMaterialize(program, context);
      if (!equalValues(result.rows, expected.in_memory_rows)) {
        failures.push(`warmup ${warmup + 1} result differs from Lisp reference`);
      }
    }
    const samples = [];
    for (let run = 0; run < runs; run += 1) {
      if (timingMode === 'gc-controlled') global.gc();
      console.log(`[js] ${expected.id} measured ${run + 1}/${runs}`);
      const result = runAndMaterialize(program, context);
      if (!equalValues(result.rows, expected.in_memory_rows)) {
        failures.push(`measured run ${run + 1} result differs from Lisp reference`);
      }
      if (structuralHash(context) !== contextSignature) {
        failures.push(`measured run ${run + 1} changed prepared context`);
      }
      samples.push({
        program_run_ms: result.runMs,
        materialize_ms: result.materializeMs,
        prepared_total_ms: result.preparedMs,
        elapsed_ms: result.preparedMs,
      });
    }
    const statistics = Object.fromEntries(
      ['program_run_ms', 'materialize_ms', 'prepared_total_ms']
        .map((phase) => [phase, phaseStats(samples.map((sample) => sample[phase]))]),
    );
    const passed = failures.length === 0;
    report.passed &&= passed;
    report.scenarios.push({
      id: expected.id,
      rows: Array.isArray(validation.rows) ? validation.rows.length : null,
      compile_ms: compileMs.get(expected.id),
      samples,
      statistics,
      context_unchanged: contextUnchanged,
      parity: { passed, failures },
      passed,
      failures,
    });
    console.log(`${passed ? 'PASS' : 'FAIL'} ${expected.id}: ${
      failures.length === 0 ? 'in-memory + PostgreSQL SQL + MariaDB SQL' : failures.join('; ')}`);
  }

  const output = option('--output', null);
  if (output !== null) {
    fs.writeFileSync(resolvePath(output), `${JSON.stringify(report, null, 2)}\n`);
  }
  console.log(`JS scale parity: ${report.scenarios.filter((s) => s.passed).length}/${report.scenarios.length} passed; mode=${timingMode}; runs=${runs}; warmups=${warmups}`);
  process.exitCode = report.passed ? 0 : 1;
}

correctedMain();
